napari/_qt/qt_main_window.py

```diff
@@ -270,12 +271,19 @@ def resize(self, width, height):
         self._qt_window.resize(width, height)
 
     def show(self):
-        """Resize, show, and bring forward the window.
-        """
+        """Resize, show, and bring forward the window."""
         self._qt_window.resize(self._qt_window.layout().sizeHint())
         self._qt_window.show()
         # make sure window is not hidden, e.g. by browser window in Jupyter
-        self._qt_window.raise_()
+        # if the application was instantiated in gui_qt(), app_name will be
+        # "napari".  If it is the second time a _qt_window has been created,
+        # then isActiveWindow() will be True (otherwise false).
+        # this is the only combination that seems to fix #732
+        app_name = QApplication.instance().applicationName()
+        if app_name == 'napari' and not self._qt_window.isActiveWindow():
+            return
+        self._qt_window.raise_()  # for MacOS
+        self._qt_window.activateWindow()  # for Windows
```

Comment:
(a) can you clarify a bit when, in *usage* terms, the window will be brought forward? It's unclear to me.

(b) I prefer logic that executes code when passing, rather than logic that bails out of the function (kinda like a double negative). So I suggest:

```python
# raise when the window is launched from Jupyter or...
if app_name != 'napari' or self._qt_window.isActiveWindow():
    self._qt_window.raise_()  # for macOS
    self._qt_window.activateWindow()  # for Windows
```

Reply:
(let me know if you wanted this in comments, as opposed to here).
There is a lot of mystery still going on here around the Qt windowing system and how it plays with the OS, and I really can't say that I understand why this works, but I think the best i know how to put it in terms of usage is as I put it above:
> `raise_()` will be called in pretty much every case except when instantiating a viewer within a `gui_qt()` context manager for the first time within the Qt app's lifecycle.

In terms of the user experience, the aim of this PR  is to never have a newly instantiated viewer window appear below any other window (i.e. below the terminal when i IPython or the browser when in a jupyter notebook).  In every case, calling `raise_()` accomplishes that; however when you `raise_` a `main_window` _immediately_ after `QApplication()` instantiation (such as in the `gui_qt()` context), for whatever reason, the menubar becomes inactive.  It's almost like the menubar belongs to a different OS process.  I feel like this is some sort of Qt race condition bug (given the prevalence of the observation around the internet).

So after some guess & check, the logic I settled on here just avoids calling `raise_` whenever the Qt App was created in the `gui_qt()` context (where we give the `QApplication().instance()` the name 'napari') and when it is the first time a viewer is being instantiated that session (in which case `_qt_window.isActiveWindow()` will be `False`).  _why_ is it false the second+ time around?  that was the guess and check part!

sorry I don't have a better explanation! :/

Reply:
@tlambert03 

> is as I put it above

This is precisely how I would like it to appear in the comment! It's useful to know what the *intent* of the if-clause is, rather than just the mechanics of it.

I'm not positive that this is always the best behaviour — I often launch a viewer and want to immediately type new things to add to it — but I'm happy to make this the default for now, and see what people think.

Reply:
that's a fair point. terminal windows are usually small(ish), so it's quite obvious where the napari window is, and one might not want to lose terminal focus.  in jupyter however, it's not uncommon for the napari window to show up _behind_ the browser window completely, and the user wouldn't know it... which is where the inspiration for `raise_` came from in the first place. (see for example this user request https://github.com/napari/napari/issues/735)

I think the name of the `QApplication().instance()` _might_ be different in those cases... would you want me to set it to _not_ raise in ipython? (fwiw, the current behavior for the last month or so has been to always raise)

Reply:
> would you want me to set it to not raise in ipython

I have a slight preference for this, but I wouldn't stress too much about it.

> fwiw, the current behavior for the last month or so has been to always raise

My daily workhorse is Linux + i3, which is a tiling window manager, so I don't really notice these things that much. =P But I do notice them when I switch to my mac for whatever reason!

Reply:
I lean towards how it will be after this PR, which will be a significant improvement for me. I'd keep the IPython and Jupyter the same for now, but we can revisit in follow-up PRs if it becomes clear that they should be treated differently

Reply:
I think I'm with @sofroniewn, I have a slight preference for raising the viewer when `napari.Viewer()` is called.  But since this seems kinda like something a user preference that might change in different circumstances, maybe we should just add a "raise" parameter to `Window.__init__`?

Currently, it looks like while the main `Window` class [does have a `show=True` parameter](https://github.com/napari/napari/blob/fffd2f2a4090f773f2aec904bee9ac28a4b1890d/napari/_qt/qt_main_window.py#L54), it's sort of impossible to access, since most people will be instantiating the window indirectly in `napari.Viewer()` [here](https://github.com/napari/napari/blob/fffd2f2a4090f773f2aec904bee9ac28a4b1890d/napari/viewer.py#L62).  Should we consider exposing some of those options a little more?  To capture all of the subtleties, we'd really need to expose both `show` and `raise_` ... since someone might want to show but not raise (like @jni here).

Reply:
hmm agreed the `show` parameter doesn't do much right now without being exposed on our top level `Viewer` / `view_*` and agreed we'd need to expose a `raise` too. I'm open to exposing, but I'm also not sure its needed (i.e. I'm not sure if anyone right now wants to use `show=False`). My preference would be to merge this PR now as is and then consider expose those additional keyword args in followup PRs

Reply:
yeah, I agree probably no need for `show=False`, and I'm fine to merge as is...
but maybe @jni wants to weigh in on whether he's ok with having `raise=True` as the default behavior, or if he wants a `raise=False` option

Reply:
@tlambert03 I'm totally fine with the behaviour, but my two comments above (add the usage comment from the PR, and change the logic flow) haven't been addressed yet!

Reply:
Ok great @jni! I'll wait for @tlambert03 to address those comments and then we can merge. Thanks for weighing in

Reply:
addressed the comments and changed the logic.  All set to merge once tests pass.